src/threei/src/threei.rs

```diff
@@ -770,6 +666,99 @@ fn _strlen_in_cage(src: *const u8, max_len: usize) -> Option<usize> {
     None // null terminator not found within max_len
 }
 
+///  Convert a byte address range [addr, addr+len) into page numbers.
+#[inline]
+fn _bytes_to_pages(addr: u64, len: u64) -> (u32, u32) {
+    // start_page: page index of the starting address
+    let start_page = (addr >> PAGESHIFT) as u32;
+    let end_addr   = addr.checked_add(len).expect("addr overflow in ensure_mapped");
+    // end_page: page index after the last byte (exclusive), ceil-rounded
+    let end_page   = ((end_addr + (PAGESIZE as u64 - 1)) >> PAGESHIFT) as u32; // ceil
+    (start_page, end_page)
+}
+
+/// Ensure that the given [addr, addr+len) region in a cage is mapped with
+/// the required protection flags. If an overlap exists, it will be removed
+/// and replaced by a new mapping.
+/// 
+/// ## Arguments
+/// - cage: reference to the target Cage (contains vmmap)
+/// - cage_id: ID of the cage for ownership metadata
+/// - addr: starting virtual address in user space
+/// - len: length of the region in bytes
+/// - required_prot: protection flags to enforce (e.g., PROT_READ|PROT_WRITE)
+/// - required_maxprot: maximum allowed protection flags
+/// - map_flags: mmap-style flags (e.g., MAP_PRIVATE|MAP_ANONYMOUS)
+/// - backing: type of memory backing (Anonymous, File, SharedMemory, etc.)
+/// 
+/// ## Returns
+/// - Ok(()) if the region was successfully mapped or already valid
+/// - Err(&'static str) if the request is invalid (out of bounds or prot violation)
+fn _ensure_user_range_mapped(
```

Comment:
is this still a function we want to have? I know we talked about this a bit today.